tracker/src/main/kotlin/org/dhis2/tracker/events/CreateEventUseCaseRepository.kt

```diff
@@ -0,0 +1,103 @@
+package org.dhis2.tracker.events
+
+import org.dhis2.commons.date.DateUtils
+import org.hisp.dhis.android.core.D2
+import org.hisp.dhis.android.core.arch.repositories.scope.RepositoryScope
+import org.hisp.dhis.android.core.event.EventCreateProjection
+import org.hisp.dhis.android.core.maintenance.D2Error
+import org.hisp.dhis.android.core.program.ProgramStage
+import java.util.Calendar.DAY_OF_YEAR
+import java.util.Date
+
+class CreateEventUseCaseRepository(
+    private val d2: D2,
+    private val dateUtils: DateUtils
+)
+
+{
+    fun createEvent(enrollmentUid: String?, programUid: String,  programStageUid: String?, orgUnitUid: String): Result<String> {
+       return try {
+            val stageLastDate = getStageLastDate(enrollmentUid, programStageUid)
+            val programStage = getProgramStage(programStageUid)
+            val eventUid = d2.eventModule().events().blockingAdd(
+                EventCreateProjection.builder().apply {
+                    enrollmentUid?.let { enrollment(enrollmentUid) }
+                    program(programUid)
+                    programStage(programStageUid)
+                    organisationUnit(orgUnitUid)
+                }.build(),
+            )
+            val eventDate = stageLastDate ?: when (programStage?.generatedByEnrollmentDate()) {
+                true -> getEnrollmentDate(enrollmentUid)
+                else -> getEnrollmentIncidentDate(enrollmentUid)
+                    ?: getEnrollmentDate(enrollmentUid)
+            }
+            val eventRepository = d2.eventModule().events().uid(eventUid)
+            val calendar = DateUtils.getInstance().getCalendarByDate(eventDate)
+            if (stageLastDate == null) {
+                val minDaysFromStart = getMinDaysFromStartByProgramStage(programStage)
+                calendar.add(DAY_OF_YEAR, minDaysFromStart)
+            } else {
+                calendar.add(DAY_OF_YEAR, programStage?.standardInterval() ?: 0)
+            }
+            if (programStage?.periodType() == null) {
+                eventRepository.setEventDate(calendar.time ?: dateUtils.today)
+            } else {
+                val nextAvailablePeriod = dateUtils.getNextPeriod(programStage.periodType(), calendar.time ?: dateUtils.today, 0)
+                eventRepository.setEventDate(nextAvailablePeriod)
```

Comment:
to improve readability, can you extract all the login regarding getting the event date to a method and then here just `eventRepository.setEventDate(getEventDate(...))` or something like, what do you think?

Reply:
of course! I just updated the PR